### DISCUSSION TITLE
Multiple select with AlpineJS and Livewire

### DISCUSSION THREAD

**Discussion by ju5t** [livewire/livewire#2419] [2. Show and tell]
There are a few cases where having a multipe select list (is that how you call it?!) can be useful. In our case, to manage permissions. The idea and some of the code came from the [forum](https://forum.laravel-livewire.com/t/select-and-select-multi/1120/10). So credits to andylord565 too, of course.

It's based on [Choises.js](https://github.com/jshjohnson/Choices). The component looks like this:

```blade
<div
    wire:ignore
    x-data="{ values: @entangle($attributes->wire('model')), choices: null }"
    x-init="
        choices = new Choices($refs.multiple, {
            itemSelectText: '',
            removeItems: true,
            removeItemButton: true,
        });

        for (const [value, label] of Object.entries(values)) {
            choices.setChoiceByValue(value || label)
        }

        $refs.multiple.addEventListener('change', function (event) {
            values = []
            Array.from($refs.multiple.options).forEach(function (option) {
                values.push(option.value || option.text)
            })
        })
    "
>
    <select x-ref="multiple" multiple="multiple">
        @foreach($options as $key => $option)
            <option value="{{ $key }}">{{ $option }}</option>
        @endforeach
    </select>
</div>
```

You can use it as follows:

```blade
<x-select.multiple wire:model="user.permissions" id="roles" :options="$permissions" />
```

- user.permissions: an array with all the selected values
- permissions: an array with all possible values

I hope it helps!

**Comment by angelamollina30** [comment]
Hi How can i display the default selected values? i want to try this on update

**Reply by samankassou** [reply]
may be by setting `user.permissions` in your livewire component

**Reply by lotje-kinable** [reply]
hi @angelamollina30 

did you, by any chance, solve setting the default values? 

thanks in advance!

**Reply by NikarashiHatsu** [reply]
Hi, sorry if this comment is considered a bump.
But somehow I manage on how to set the default selected value below:

```js
for (const [value, label] of Object.entries(values)) {
    choices.setChoiceByValue(`${label}`)
}
```

If I see the `values`  through the console, it is a index-value pair.
So it's considered as `for (const [index, value] of Object.entries(values))`.

It might be a different approach for a different case, in my case, it just works.

**Comment by hadesunseenn** [comment]
@ju5t it works fine until we hit submit button. So lets say I am in edit view of form and choices styles loads properly but when we hit submit the styles of choices are gone. Prolly because we are using x-init="" and when we hit submit the livewire component re-renders

**Reply by Christophvh** [reply]
anybody got a clue how to fix re-render issues? having real trouble with it.

**Reply by joshhanley** [reply]
@Christophvh can you be more specific, what render issues? Can you share a screen capture demonstrating it? The more detail you can provide, the easier it is for people to assist you 🙂

If the issue is what @hadesunseenn is describing, make sure that you both have `wire:ignore` on the wrapping `<div>` as per the example above, as that stops Livewire from clearing out the styles.

Hope this helps!

**Reply by Christophvh** [reply]
wire:ignore is pretty much a 'hack'  because you lose all functionality. I have a choicesJS dropdown to add items to a list. When I click a button 'addToList' the choices dropdown breaks because the component gets re-rendered. But I need the re-render because the dropdown options need to change dynamically based on livewire properties.

**Reply by ju5t** [reply]
I wouldn't call it a hack, `wire:ignore` is a necessity. 

I don't have much time to make it work for your use case at the moment (sorry!) but I'd probably use `$watch` to check the `values` of the Alpine object and have it reset all the options for you.

That's probably the easiest solution here. See https://alpinejs.dev/essentials/lifecycle#watch.

**Reply by joshhanley** [reply]
@Christophvh `wire:ignore` isn't a hack, it's essential when using javascript libraries that modify the DOM with Livewire. See here https://laravel-livewire.com/docs/2.x/alpine-js#ignoring-dom-changes

Yes it does mean you can't use blade to update the options. That was why Alpine was written as a layer between Livewire and any javascript libraries.

What you need to do is use Alpine to get the new options from Livewire and use the libraries javascript API to programmatically update the options.

Hope this helps!

**Reply by joshhanley** [reply]
@ju5t I should have read your whole message before responding, I just said basically the same thing 😆

**Reply by ju5t** [reply]
@joshhanley Ha, finally a time where I don't need your help with Livewire (and in this case, Alpine)! 🥲

@Christophvh the example I made is based on html. You don't have to and in this case I'd probably use a different approach. I don't have a working demo set up right now to build and verify it (we ended up not using it ourselves), but the gist of it is:

- `@entangle` a new property to the Blade-component. Call it something like `choices`. You might need to use `@js` too. There might be some caveats here that I didn't consider.
- In the init-part of the component use [setChoices](https://github.com/Choices-js/Choices#setchoiceschoices-value-label-replacechoices) to set all options.
- Then use the `$watch` helper to call `setChoices` again if the value changes.

I think this could work but it might need some fiddling on your end.

**Reply by Christophvh** [reply]
@ju5t @joshhanley thanks for the help. I understand what the issue is and that is related to the DOM diffing. 

Stackoverflow/Github is filled with people with this issue and none of the threads have a working example will full reactivity. they all stop at` wire:ignore` or just refreshing the page.

This is a very common problem, this problem happens every time for every external JS lib anybody uses. And there is no common solution for any of them. Can't find any examples to get `$watch` working with choices, only Select2. The online solution is `wire:ignore`, no solutions on how to get it re-active with watch. I appreciate all the work Livewire and AlpineJS have done and it has its place. But having to use LiveWire + AlpineJs both and fiddling around with @entangle() @js() $watch and having to figure out which is from which package is insane to teach fellow developers and has a learning curve that doesn't translate to anything else. Switching back to React, Vue is just way easier when the app gets more complicated and react knowledge translates to all kinds of apps not just the laravel eco-system.

**Reply by ju5t** [reply]
@Christophvh there are Alpine only examples available here: https://alpinejs.dev/component/choices. It's paid, but definitely worth it considering @calebporzio teaches you how he built the component. And it's not that one component. The only downside is that it's just Alpine. The components there do not include Livewire, unfortunately.

> no solutions on how to get it re-active with watch

It's like `Hello world!`; you need to start at somewhere. There's actually good documentation available for `$watch` in the Alpine documentation: https://alpinejs.dev/magics/watch. In fact, it's pretty similar and inspired by `$watch` in Vue. Like a lot of things are in Alpine.

>  Switching back to React, Vue is just way easier when the app gets more complicated

If that's what you're used to it probably is easier. We tried Vue but thought it be to too complicated for our use case. And we have a lot of reactivity in our app. We settled on Livewire and have not found anything we could not do with it.

I think both have their advantages and disadvantages, but it's all personal preference. All of them would have specifics you need to understand to get things working. There's no way around that.

**Reply by joshhanley** [reply]
@Christophvh no worries! 🙂

Yeah it certainly can seem a lot, sadly though, it's the nature of DOM diffing and external libraries that modify the DOM (a problem not unique to Livewire). While this is a common problem, there is no common solution for it as every library works differently.

Personally I didn't have much issue with when I was learning it all, as the only time I had to reach for an integration like this was to build a working component. Once the component was running, then I just consumed it in my app. That being said though, I generally don't use other libraries, like choices, if possible, as I prefer to build native components using Alpine/ Livewire without the third party dependency.

Livewire's community is still young too, so overtime, people will make and share more components, both native and integrations with third parties, which will make it easier for everyone to just consume them. But for now we just have to make them ourselves or take someone else's solution and expand on it, and hopefully share it with others(like @ju5t has done here).

Vue and React components are aplenty these days, but I remember when Vue first came out, it didn't integrate with other libraries very well either, but now has a thriving ecosystem, with packages for any type of component we may need.

It all comes down to personal preference, if Vue or React is easier/ preferable, then use that for you app. My preference used to be Vue but now is Livewire and Alpine and I don't use anything else. Use whatever makes you most productive 🙂

That being said, we're happy to try and assist as much as we can. So if you do decide to try and add reactivity to the choices component, and get stuck, feel free to post up and we can see if we can assist 🙂

We also have a discord channel https://discord.gg/livewire which can be better for assisting with things like this as you get more instant back and forth.

Hope this helps!

**Comment by marksparrish** [comment]
I am not getting the array back into user.permissions. In my case it is _'orgIds'_. It keeps coming back as "null" and never validates.

My livewire component
```
<?php 

namespace App\Http\Livewire\Admin\Users;

use App\Models\Organization as Model;
use Livewire\Component;

class AddOrganization extends Component
{
    public $roles;
    public $availableOrgs;
    public $user;
    public $orgIds;
    public $orgRole;

    protected $rules = [
        'orgIds' => ['required'],
        'orgRole' => ['required'],
    ];

    public function mount()
    {
        $this->roles = Model::ORGANIZATION_ROLES;
        $this->availableOrgs = Model::whereNotIn('id', $this->user->organizations()->pluck('id')->toArray())->get();
    }

    public function add()
    {
        dd($this->orgIds, $this->orgRole);
        $validatedData = $this->validate();
        dd($validatedData);
    }

    public function render()
    {
        return view('livewire.admin.users.add-organization');
    }
```
My blade file
```
<div>
    <div class="card">
        <div class="p-1 card-body">
            <div class="flex justify-between">
                <h2 class="card-title text-lg font-semibold">
                    <span class="hidden sm:block">Organization Memberships</span>
                    <span class="sm:hidden">Organizations</span>
                </h2>
                <div class="card-actions">
                    <label for="add-org-modal" class="btn btn-success btn-sm modal-button">
                        <span class="hidden sm:block">Join Organizations</span>
                        <span class="sm:hidden">
                            <x-icon.plus />
                        </span>
                    </label>
                </div>
            </div>
            <table class="table w-full">
                <thead>
                    <tr>
                        <th>Name</th>
                        <th>Role</th>
                        <th>
                            <span class="sr-only">Leave</span>
                        </th>
                    </tr>
                </thead>
                <tbody class="divide-y divide-gray-200 bg-white">
                    @forelse ($user->organizations as $model)
                    <tr>
                        <td>{{ $model->name }}</td>
                        <td>{{ Illuminate\Support\Str::ucfirst($model->pivot->role) }}</td>
                        <td>
                            <button type="button" wire:click="remove({{ $model->id }})" class="btn btn-sm btn-error">-</button>
                        </td>
                    </tr>
                    @empty
                    <tr>
                        <td colspan="3">No Assigned Organizations</td>
                    </tr>
                    @endforelse
                </tbody>
            </table>
        </div>
    </div>

    <div>
        <form wire:submit.prevent="add">
            <input type="checkbox" id="add-org-modal" class="modal-toggle" />
            <label for="add-org-modal" class="modal cursor-pointer">
                <label class="modal-box relative" for="">
                    <div class="flex flex-col space-y-2">
                        <div class="text-lg font-bold">Add User to Organization</div>
                        <div class="divider"></div>
                        <div class="space-y-2">
                            <div class="form-control">
                                <p class="text-base font-semibold">Select Users Role</p>
                                @foreach($roles as $role)
                                <div class="ml-4 mb-2">
                                    <input type="radio" wire:model="orgRole" value="{{ $role }}" class="radio" />
                                    <span class="uppercase ml-2">{{ $role }}</span>
                                </div>
                                @endforeach
                                @error('orgRole')
                                <x-form.input-error>
                                    {{ $message }}
                                </x-form.input-error></span>
                                @enderror
                            </div>
                            <div class="form-control w-full max-w-xs">
                                <p class="text-base pb-2 font-semibold">Select Organization(s)</p>
                                <x-select-box wire:model="orgIds" id="organizations" :options="$availableOrgs" />
                                @error('orgIds')
                                <x-form.input-error>
                                    {{ $message }}
                                </x-form.input-error></span>
                                @enderror
                            </div>

                        </div>
                        <div class="btn-group justify-end">
                            <div class="btn-group grid grid-cols-2">
                                <button class="btn btn-sm btn-error"><label for="add-org-modal">Cancel</label></button>
                                <button type="submit" class="btn btn-sm btn-success">Join Org</button>
                            </div>
                        </div>
                    </div>
                </label>
            </label>
        </form>
    </div>
</div>
```
The multiselect blade component
```
<div
    wire:ignore
    x-data="{ values: @entangle($attributes->wire('model')), choices: null }"
    x-init="
        choices = new Choices($refs.multiple, {
            itemSelectText: '',
            removeItems: true,
            removeItemButton: true,
        });

        for (const [value, label] of Object.entries(values)) {
            choices.setChoiceByValue(value || label)
        }

        $refs.multiple.addEventListener('change', function (event) {
            values = []
            Array.from($refs.multiple.options).forEach(function (option) {
                values.push(option.value || option.text)
            })
        })
    "
>
    <select x-ref="multiple" multiple="multiple">
        @foreach($options as $key => $option)
            <option value="{{ $key }}">{{ $option }}</option>
        @endforeach
    </select>
</div>
```

I call the blade component -
`<x-select-box wire:model="orgIds" id="organizations" :options="$availableOrgs" />`

**Reply by marksparrish** [reply]
If I forward attributes - 
```
        <select {{ $attributes }} x-ref="multiple" multiple="multiple">
            @foreach($options as $key => $option)
            <option value="{{ $key }}">{{ $option }}</option>
            @endforeach
        </select>
```
I can get an array but it only has the last value chosen.

**Reply by marksparrish** [reply]
> If I forward attributes -
> 
> ```
>         <select {{ $attributes }} x-ref="multiple" multiple="multiple">
>             @foreach($options as $key => $option)
>             <option value="{{ $key }}">{{ $option }}</option>
>             @endforeach
>         </select>
> ```
> 
> I can get an array but it only has the last value chosen.

Ok - disregard. This is an anonymous component.
